numpy/__init__.pxd

```diff
@@ -1018,6 +1022,76 @@ cdef extern from *:
     """
 
 
+cdef inline bint is_integer_object(object obj):
+    """
+    Cython equivalent of
+
+    `isinstance(val, (int, long, np.integer)) and not isinstance(val, bool)`
+
+    Parameters
+    ----------
+    val : object
+
+    Returns
+    -------
+    is_integer : bool
+
+    Notes
+    -----
+    This does not counts np.timedelta64 objects as integers.
+    """
+    return (not isinstance(obj, bool) and PyArray_IsIntegerScalar(obj)
+            and not is_timedelta64_object(obj))
+
+
+cdef inline bint is_float_object(object obj):
+    """
+    Cython equivalent of `isinstance(val, (float, np.complex_))`
+
+    Parameters
+    ----------
+    val : object
+
+    Returns
+    -------
+    is_float : bool
+    """
+    return (isinstance(obj, float) or
+            (PyObject_TypeCheck(obj, &PyFloatingArrType_Type)))
+
+
+cdef inline bint is_complex_object(object obj):
+    """
+    Cython equivalent of `isinstance(val, (complex, np.complex_))`
+
+    Parameters
+    ----------
+    val : object
+
+    Returns
+    -------
+    is_complex : bool
+    """
+    return (isinstance(obj, complex) or
+            PyObject_TypeCheck(obj, &PyComplexFloatingArrType_Type))
+
+
+cdef inline bint is_bool_object(object obj):
+    """
+    Cython equivalent of `isinstance(val, (bool, np.bool_))`
```

Comment:
Does cython not just let you run this directly?

Reply:
it does, but it doesnt optimize it into a C call

Reply:
Is the optimization even worth it? Isn't `isinstance` a thin wrapper around `PyType_Check` anyway?

Would it be worth filing a issue against cython requesting for it to optimize this type of thing automatically?

These just seem like weird helper functions to add to me, and I the only justification I can think of would be upstream deficiencies - in which case I'd like to see links to open cython issues in our source code.

Reply:
There is a difference, `PyObject_TypeCheck` does not support metaclasses/`__subclasscheck__` or `__subclasshook__`, so for this type of function it may actually make a real difference.

Reply:
... except that the function calls `isinstance(obj, bool)` in the first half

Reply:
Would cython allow the user to use `PyObject_TypeCheck(obj, np.bool_)` instead of referring to the c type directly?

Reply:
Fair point Matti, I thought that maybe cython manages to optimize it for some basic types, but it does not... In which case I am a bit confused though, the tuple version, is identical from the C-side, no? It both translates 1:1 to `PyObject_IsInstance` check, so it would seem at best this achieves a half optimization of the full thing?

@eric-wieser I suspect you may need to do `<PyTypeObject *>np._bool`.

Reply:
If this optimization really matters, can we instead persuade `cython` to add a builtin `isinstance_fast(x, (t1, t2))`, which skips just the hooks? The alternative sounds like it's to add a `is_X` function for every type `X`, which seems ridiculous to me.

Reply:
It is a micro-optimization, whether it matters in real life, I am not sure. The nature of this type of optimization is that it is may be clear in `%timeit` but in any real program might just swept away by other things.

I don't know if cython can easily do `isinstance_fast(x, (t1, t2))` mapping to `PyObject_TypeCheck(x, t1) || PyObject_TypeCheck(x, t2)`, but I suppose exposing `PyObject_TypeCheck` itself in Cython could be nice... (For bool, it actually can be optimized to `type(obj) is bool`, which cython already supports :)).

Reply:
> (For bool, it actually can be optimized to `type(obj) is bool`, which cython already supports :))

That optimization would be incorrect though, for `class Foo(bool): pass`?

Reply:
Its not allowed to subclass bools.

Reply:
(btw. for some other built-ins, It actually also makes sense to use the `PyList_Check`, etc. because Python has a super-fast path for those optimizing them into simple flag checks, the type object flags have flags for to signal subclass of list, float, tuple, and some others)

Reply:
> for some other built-ins, It actually also makes sense to use the PyList_Check

cython makes this substitution automatically.  similar for PyFloat_Check and most other basics

Reply:
Oh, I tried with bool, it seems Cython missed that it makes sense to use `PyBool_Check` as well, probably because it does not use the flags, but rather the fact that its not subclassable.

Reply:
> except that the function calls isinstance(obj, bool) in the first half

cython optimizes this to PyBool_Check

> Is the optimization even worth it? Isn't isinstance a thin wrapper around PyType_Check anyway?

it has to go into python-space to do a module-level lookup of `np`, followed by `PyObjectGetAttr(np, "dtype")` followed by ...

> can we instead persuade cython to add a builtin isinstance_fast(x, (t1, t2)), which skips just the hooks? The alternative sounds like it's to add a is_X function for every type X, which seems ridiculous to me.

in my experience the cython maintainers have been very helpful about allowing optimizations, but i have enough trouble grokking their code that i've never been able to implement any of those optimizations myself.

> These just seem like weird helper functions to add to me

for e.g. `ndarray` i can do `from numpy cimport ndarray` and then `isinstance(obj, ndarray)` checks are much faster than `isinstance(obj, np.ndarray)`, though IIRC still slower than `PyArray_Check(obj)`.  if analogues were exposed for datetime64/timedelta64/etc, that might be a cleaner solution.

> It is a micro-optimization, whether it matters in real life, I am not sure. The nature of this type of optimization is that it is may be clear in %timeit but in any real program might just swept away by other things.

These are getting called in pandas' low-level code and often in loops, so it definitely makes a difference.

Reply:
Woops, past-me knew that PyBoolCheck is one of the handful they dont get: https://github.com/cython/cython/issues/2690

Reply:
Anyway, you are right,its worthwhile to do typechecks like this. The only real question is probably if we want to expose this in NumPy directly and if with which name. As such these are pretty much no-maintenance needed functions at least...

Reply:
Would a sub-namespace of `cnp` make sense for this type of helper? Does cython make it easy to create such a namespace?

Reply:
> Does cython make it easy to create such a namespace?

Yes, but it works slightly different from python namespacing.

ATM cython distributes a `numpy/__init__.pxd` and a `numpy/math.pxd`, so let's use these two for exposition.

In my .pyx file:

```
cimport numpy as cnp
cimport numpy.math as npy_math

def foo(x):  # <-- works
    return cnp.whatever(x)

def bar(x):  # <-- works
    return npy_math.thing(x)

def baz(x):  # <-- doesnt work
    return cnp.math.thing(x)
```

To make `baz` work, the `__init__.pxd` file would also need to have a `from . cimport math`, which it currently does not.

Reply:
So the only difference from python is that `cimport x.y` does not perform a `setattr(x, 'y', mod_obj)`?

Reply:
AFAIU

Reply:
So to summarize so far:
- this optimization is worth it, but maybe belongs upsrtream in Cython, which doesn't seem easy to do
- there was a suggestion to put these into a namespace, which seems non-trivial

which leaves us with accepting these functions as work-arounds, much like the python work-arounds we have in other places. Perhaps we can mark them as such with some comments and accept this as-is?

Reply:
> this optimization is worth it, but maybe belongs upsrtream in Cython, which doesn't seem easy to do

not sure i understand this.  My plan is that if/when this is merged, I'll make a PR with cython to make the same edits to their `numpy/__init__.pxd` file

Reply:
These functions to two things: First they do some optimizations that cython should do in `isinstance()`, but fails to do. Second, they make writing `isinstance(value, (python_type, numpy_type))` slight shorter.  I think what Matti is saying that the second point may not be worth it.  But the first point is worth it until cython is capable of parsing the full `isinstance` check into something faster (or maybe providing an `isinstancefast`/`hastype` check with that signature).  Once Cython has such a new feature, these functions could probably be considered deprecated.

Reply:
Cython does translate `isinstance(value, (python_type, numpy_type))` into the corresponding C-API calls, except for `bool`, as @jbrockmendel pointed out. Calling `isinstance(obj, (float, floating))` should be just fine.